domain/permission/state/shared.go

```diff
@@ -0,0 +1,79 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+
+	"github.com/canonical/sqlair"
+	"github.com/juju/errors"
+
+	corepermission "github.com/juju/juju/core/permission"
+	permissionerrors "github.com/juju/juju/domain/permission/errors"
+	internaldatabase "github.com/juju/juju/internal/database"
+)
+
+// SharedState holds the shared state for the permission domain. Methods
+// which are shared between multiple domains.
+type SharedState struct {
+	internaldatabase.StatementBase
+}
+
+// NewSharedState creates a new SharedState.
+func NewSharedState(base internaldatabase.StatementBase) *SharedState {
+	return &SharedState{StatementBase: base}
+}
+
+// AddUserPermissionArgs is a specification for adding a user permission.
+type AddUserPermissionArgs struct {
+	PermissionUUID string
+	UserUUID       string
+	Access         corepermission.Access
+	Target         corepermission.ID
+}
+
+// AddUserPermission adds a permission for the given user on the given target.
+// TODO (stickupkid): Work out if there is a better location for common
+// state functions.
+func (s SharedState) AddUserPermission(ctx context.Context, tx *sqlair.TX, spec AddUserPermissionArgs) error {
+	// Insert a permission doc with
+	// * permissionObjectAccess as permission_type_id
+	// * uuid of the user (spec.User) as grant_to
+	// * spec.Target.Key as grant_on
+	newPermission := `
+INSERT INTO permission (uuid, permission_type_id, grant_on, grant_to)
+VALUES ($addUserPermission.uuid, $addUserPermission.permission_type_id, $addUserPermission.grant_on, $addUserPermission.grant_to)
```

Comment:
Surely something here could be a *, I think @Aflynn50 said that he added support for INSERT statements?

Reply:
I'll make a note with @Aflynn50 to do a followup PR.